### PR TITLE
WIP: Enhancing VPR lookahead map

### PR DESCRIPTION
**Pull request by acomodi**
Signed-off-by: Alessandro Comodi <acomodi@antmicro.com>

#### Description

This PR is an initial Draft with the effort of getting the lookahead map work well with the Series 7 architectures.

#### Related Issue

The issue tracking what needs to be done and all suggestions is here: https://github.com/verilog-to-routing/vtr-verilog-to-routing/issues/1325

#### Motivation and Context

Currently, the SymbiFlow VTR fork implements the [connection box lookahead map](https://github.com/SymbiFlow/vtr-verilog-to-routing/tree/wip/lookahead_sampling2) which has been tested to work well with Series 7 devices.
The idea is to get the lookahead map code produce similar results, so that eventually, there can be only one lookahead code being used.

In case the two different approaches cannot be merged, an alternative would be to have a common utility library for the different lookaheads.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by acomodi** [comment]
@litghost, @kmurray, @HackerFoo FYI.

I have completed the merge of the connection box lookahead map and the upstream lookahead map.

From the connection box lookahead, the following has been taken:
- Sampling method: this has been extracted in a separate file
- Paths expansion and lookahead filling
- Lookahead map format (different capnp schema)

From the upstream lookahead map:
- OPIN --> CHAN information
- CHAN --> IPIN information
- Delta locations calculations
- Map addressability: e.g. `cost_map[CHAN][SEG_ID][DELTA_X][DELTA_Y]`

There are two assertions that needed to be skipped as they were triggered during the place delay matrix calculations:
- https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/1b9065116a6939eea16204885d9d53997a646186/vpr/src/place/timing_place.h#L196
- https://github.com/SymbiFlow/vtr-verilog-to-routing/blob/8980e46218542888fac879961b13aa7b0fba8432/vpr/src/route/connection_box_lookahead_map.cpp#L488

The last one is due to the fact that there are some non valid values for the expected cost which is set to infinite.

I also had to temporarily remove the lookahead generation test under `vpr/test` which still needs to be adapted to the new lookahead generation mechanism.

With this implementation, the routing step has the following results:
- 238.30 seconds
- 18.796 ns (CPD)
- Works on HW

With the connection box lookahead map:
- 146.26 seconds
- 17.381 ns (CPD)
- Works on HW

There is an increase in the CPD which I still need to analyse, as well as an increased run-time. Both of them are probably due to mispredictions, which I still need to analyse in details.

**Comment by litghost** [comment]
> There are two assertions that needed to be skipped as they were triggered during the place delay matrix calculations:

Both of these assertions should be in place, and violating them likely indicates a bug?

File: libs/libvtrcapnproto/CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,7 @@ endif()
 # Each schema used should appear here.
 set(CAPNP_DEFS
     place_delay_model.capnp
+    connection_map.capnp
```
Comment (litghost) [inline libs/libvtrcapnproto/CMakeLists.txt:22]: connection_map.capnp shouldn't be required, because it isn't used?

File: vpr/src/route/route_profiling.cpp [hunk 2]
```diff
@@ -29,6 +30,10 @@ void time_on_fanout_analysis() {}
 
 void profiling_initialization(unsigned /*max_net_fanout*/) {}
 
+void conn_start() {}
```
Comment (litghost) [inline vpr/src/route/route_profiling.cpp:33]: The changes to route_profiling.cpp should likely be in another PR.

File: vpr/src/route/route_timing.cpp [hunk 3]
```diff
@@ -1096,6 +1096,8 @@ bool timing_driven_route_net(ConnectionRouter& router,
             conn_delay_budget.short_path_criticality = budgeting_inf.get_crit_short_path(net_id, target_pin);
         }
 
+        profiling::conn_start();
```
Comment (litghost) [inline vpr/src/route/route_timing.cpp:1099]: Should move out of this PR with the route_profiling.h/.cpp changes

File: libs/libvtrcapnproto/CMakeLists.txt (hunk 1 above)
Comment (litghost) [reply libs/libvtrcapnproto/CMakeLists.txt:22]: Or if we are going to use this, it should move to map_lookahead.capnp

File: vpr/test/test_map_lookahead_serdes.cpp [hunk 4]
```diff
@@ -1,62 +0,0 @@
-#include "catch.hpp"
```
Comment (litghost) [inline vpr/test/test_map_lookahead_serdes.cpp:1]: Why did this test get deleted?

File: vpr/src/route/rr_graph_storage.h [hunk 5]
```diff
@@ -48,7 +48,7 @@
  *       This field is valid only for IPINs and OPINs and should be ignored  *
  *       otherwise.                                                          */
 struct alignas(16) t_rr_node_data {
-    int8_t cost_index_ = -1;
+    int16_t cost_index_ = -1;
```
Comment (litghost) [inline vpr/src/route/rr_graph_storage.h:51]: This should be put in another PR, to examine changing this independently of the other changes.

File: vpr/src/route/router_lookahead_map.h [hunk 6]
```diff
@@ -1,48 +1,84 @@
-#pragma once
+#ifndef CONNECTION_BOX_LOOKAHEAD_H_
+#define CONNECTION_BOX_LOOKAHEAD_H_
```
Comment (litghost) [inline vpr/src/route/router_lookahead_map.h:2]: Either leave this as #pragma once, or fix the #define

File: vpr/test/test_map_lookahead_serdes.cpp (hunk 4 above)
Comment (acomodi) [reply vpr/test/test_map_lookahead_serdes.cpp:1]: This was based on the previous router_lookahead_map code, I would need to readapt the test with the code merged from the connection box lookahead map and the upstream lookahead map. It still is in the TODO list.

**Comment by litghost** [comment]
This PR likely combines too many changes at once.  Things unrelated to the lookahead change should go in their own PR, and the initial change to the map lookahead should be extensions that do not change behavior significantly.  For example, the CHAN to IPIN PR should be separated from the sampling / parallelism PR.

**Comment by acomodi** [comment]
@litghost All right, I'll split this into multiple ones in the following order to have an incremental merge of the changes:

- Route profiling (this should be independent from the other PRs)
- Sampling method
- Lookahead generation
- OPIN --> CHAN and CHAN --> IPIN

**Comment by kmurray** [comment]
>  I'll split this into multiple ones in the following order to have an incremental merge of the changes

I agree this seems like the right approach.

**Comment by acomodi** [comment]
Closing, superseded by https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1449
